scapy/arch/windows/__init__.py

```diff
@@ -67,16 +69,58 @@ def _encapsulate_admin(cmd):
     # To get admin access, we start a new powershell instance with admin rights, which will execute the command
     return "Start-Process PowerShell -windowstyle hidden -Wait -Verb RunAs -ArgumentList '-command &{%s}'" % cmd
 
+class _PowershellManager(Thread):
+    """Used only when scapy boots. Instance used to send multiple
+    commands on the same Powershell process. Skips the powershell
+    booting time to speed up scapy's boot"""
+    def __init__(self):
+        self.process = sp.Popen([conf.prog.powershell, "-NoLogo", "-NonInteractive", "-Command", "-"], # Start & redirect input
+                         stdout=sp.PIPE,
+                         stdin=sp.PIPE,
+                         stderr=sp.STDOUT,
+                         universal_newlines=True)
+        self.buffer = []
+        self.running = True
+        Thread.__init__(self)
+        self.start()
+
+    def run(self):
+        while self.running:
+            self.buffer.append(self.process.stdout.readline())
+
+    def query(self, cmd):
+        self.buffer = []
+        query = cmd + "; echo scapy_end\n"
+        self.process.stdin.write(query)
+        self.process.stdin.flush()
+        while self.running:
```

Comment:
That's actually a busy loop.
I wonder if using any synchronization mechanism (event, maybe) wouldn't bring advantage.
Like query doing self.event.wait() and run() detecting "scapy_end" and calling self.event.set().
But without profiling I don't know if it's worth the effort.

(All the comments here coming from the fact that in this implementation query is the longest code to execute, so might be worth optimizing)

Reply:
(Yes, Get-NetRoute is not supported in win7, I'm talking about executing whole netsh command)

Reply:
Well...
It *made* a difference. I went down from 4.9 s to 3.6 when profiling the code in the same conditions.

Here comes the diff, sorry for the format, I didn't have time to format it properly:

```
diff -r scapy/arch/windows/__init__.py /c/Python27/Lib/site-packages/scapy/arch/windows/__init__.py
16c16
< from threading import Thread
---
> from threading import Thread, Event
84a85
>         self.event = Event()
89c90,94
<             self.buffer.append(self.process.stdout.readline())
---
>             read_line = self.process.stdout.readline()
>             if "scapy_end" in read_line:
>                 self.event.set()
>             else:
>                 self.buffer.append(read_line)
92d96
<         self.buffer = []
96,100c100,101
<         while self.running:
<             if self.buffer and len(self.buffer) >= 1:
<                 if "scapy_end\n" in self.buffer:
<                     break
<         self.buffer.remove("scapy_end\n")
---
>         self.event.wait()
>         self.event.clear()
757,758c758,766
<     ps = sp.Popen([conf.prog.cmd, '/c', 'netsh interface ipv6 show route level=verbose'], stdout = sp.PIPE)
<     stdout, stdin = ps.communicate()
---
>     query_cmd = "netsh interface ipv6 show route level=verbose"
>     if POWERSHELL_PROCESS.running:
>         stdout = POWERSHELL_PROCESS.query(query_cmd)
>     else:
>         stdout = sp.Popen([conf.prog.powershell] + query_cmd,
>                    stdout=sp.PIPE,
>                    universal_newlines=True).stdout
>         # ps = sp.Popen([conf.prog.cmd, '/c', 'netsh interface ipv6 show route level=verbose'], stdout = sp.PIPE)
>         stdout, stdin = ps.communicate()
```

Reply:
Alright, will do

Reply:
A few things: scapy_end is contained in the first line (the command gets printed on some versions), universal_newlines will crash on python 3